breacharbiter_test.go

```diff
@@ -0,0 +1,782 @@
+	},
+}
+
+// TestMockRetributionStore instantiates a mockRetributionStore and tests its
+// behavior using the general RetributionStore test suite.
+func TestMockRetributionStore(t *testing.T) {
+	for _, test := range retributionStoreTestSuite {
+		t.Run(
+			"mockRetributionStore."+test.name,
+			func(tt *testing.T) {
+				mrs := newMockRetributionStore()
+				frs := newFailingRetributionStore(
+					func() RetributionStore { return mrs },
+				)
+				test.test(frs, tt)
+			},
+		)
+	}
+}
+
+// TestChannelDBRetributionStore instantiates a retributionStore backed by a
+// channeldb.DB, and tests its behavior using the general RetributionStore test
+// suite.
+func TestChannelDBRetributionStore(t *testing.T) {
+	// First, create a temporary directory to be used for the duration of
+	// this test.
+	tempDirName, err := ioutil.TempDir("", "channeldb")
+	if err != nil {
+		t.Fatalf("unable to initialize temp "+
+			"directory for channeldb: %v", err)
+	}
+	defer os.RemoveAll(tempDirName)
+
+	// Disable logging to prevent panics bc. of global state
+	channeldb.UseLogger(btclog.Disabled)
+
+	// Next, create channeldb for the first time.
+	db, err := channeldb.Open(tempDirName)
+	if err != nil {
+		t.Fatalf("unable to open channeldb: %v", err)
+	}
+	defer db.Close()
+
+	restartDb := func() RetributionStore {
+		// Close and reopen channeldb
+		if err = db.Close(); err != nil {
+			t.Fatalf("unalbe to close channeldb during restart: %v",
+				err)
+		}
+		db, err = channeldb.Open(tempDirName)
+		if err != nil {
+			t.Fatalf("unable to open channeldb: %v", err)
+		}
+
+		return newRetributionStore(db)
+	}
+
+	// Finally, instantiate retribution store and execute RetributionStore
+	// test suite.
+	for _, test := range retributionStoreTestSuite {
+		t.Run(
+			"channeldbDBRetributionStore."+test.name,
+			func(tt *testing.T) {
+				if err = db.Wipe(); err != nil {
+					t.Fatalf("unable to wipe channeldb: %v",
+						err)
+				}
+
+				frs := newFailingRetributionStore(restartDb)
+				test.test(frs, tt)
+			},
+		)
+	}
+}
+
+// countRetributions uses a retribution store's ForAll to count the number of
+// elements emitted from the store.
+func countRetributions(t *testing.T, rs RetributionStore) int {
+	count := 0
+	err := rs.ForAll(func(_ *retributionInfo) error {
+		count++
+		return nil
+	})
+	if err != nil {
+		t.Fatalf("unable to list retributions in db: %v", err)
+	}
+	return count
+}
+
+// testRetributionStore executes a generic test suite for any concrete
+// implementation of the RetributionStore interface.
+func testRetributionStoreAddRemove(frs FailingRetributionStore, t *testing.T) {
+	// Make sure that a new retribution store is actually emtpy.
+	if count := countRetributions(t, frs); count != 0 {
+		t.Fatalf("expected 0 retributions, found %v", count)
+	}
+
+	// Add all retributions, check that ForAll returns the correct
+	// information, and then remove all retributions.
+	testRetributionStoreAdds(frs, t, false)
```

Comment:
We may want to consider making these test cases into proper [sub-tests](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks). As this way, it'll be easy to isolate a particular case for execution within the wider test case.